### ISSUE TITLE
Integration not compatible with Autofac 5.0 Release Candidate

### ISSUE THREAD

**Issue by alistairjevans** [autofac/Autofac.Mvc#28]
The Autofac.Mvc integration has build errors when built against the latest Autofac Release candidate in MyGet, specifically:

```
src\Autofac.Integration.Mvc\RegistrationExtensions.cs(457,44,457,60): error CS1501: No overload for method 'ResolveComponent' takes 2 arguments
```

The IComponentContext.ResolveComponent method has changed signature, which breaks one specific bit of code.

In addition, I believe the NuSpec will need updating to change the version constraints from ``[3.5.0,5.0.0)`` to ``[5.0.0, 6.0.0)``, because this version will no longer be compatible with older Autofac versions.

**Comment by tillig** [comment]
That sounds right to me.

**Comment by alistairjevans** [comment]
I would appreciate some guidance (perhaps from @alexmg) on how to correctly provide a ``Service`` into a ``ResolveRequest`` constructor, or get access to the correct ``ResolveRequest``, inside some of the more fruity registrations, such as the one causing the problem in this case (full code is [here](https://github.com/autofac/Autofac.Mvc/blob/develop/src/Autofac.Integration.Mvc/RegistrationExtensions.cs#L431-L467)):

```
.OnRegistered(e => e.ComponentRegistry.Register(RegistrationBuilder
    .ForDelegate((c, p) =>
    {
        var session = HttpContext.Current.Session;
        object result;
        lock (session.SyncRoot)
        {
            result = session[e.ComponentRegistration.Id.ToString()];
            if (result == null)
            {
                // Need a ResolveRequest here
                result = c.ResolveComponent(e.ComponentRegistration, p);
                session[e.ComponentRegistration.Id.ToString()] = result;
            }
        }
        return result;
    })
    .As(services)
    .InstancePerLifetimeScope()
    .ExternallyOwned()
    .CreateRegistration()));
```

If the purpose of the ``Service`` in the ``ResolveRequest`` is to know which service was originally use to resolve a registration, what do I use when there is a list of services and the delegate doesn't pass through the requested service into the delegate? 

Should the delegate call have a new overload in core that allows the delegate to take the requested service?

It might be that I'm missing something.

**Comment by alexmg** [comment]
The intent of the `Service` on the `ResolveRequest` is to know which of the multiple possible services assigned to a registration was actually being requested, and primarily for our own internal benefit. In regular usage this is almost always an explicit decision made by the user: they either resolve the service directly from an `IComponentContext` or have indicated the desired service by declaring it as a dependency in a constructor. It is only in some of these infrastructure type scenarios where you might want to perform a resolution based on a registration instead of a specific service.

In this particular case the original registration is being hijacked and left orphaned from direct access because its services have been cleared (clever but a bit of a hack). This code is not replacing the original registration but is actually making it so that the registration it adds is the only externally available mechanism to get an instance of the original registration. Because the `Id` of the original registration is used to as the key for the session cache it would be possible to loop around the original services and create a registration for each. When resolved, any of those would point back to the original registration, and would also retrieve the same instance from the cache once already activated.

```
return registration
    .ExternallyOwned()
    .OnRegistered(e =>
    {
        foreach (var service in services)
        {
            e.ComponentRegistry.Register(RegistrationBuilder
                .ForDelegate((c, p) =>
                {
                    var session = HttpContext.Current.Session;
                    object result;
                    lock (session.SyncRoot)
                    {
                        result = session[e.ComponentRegistration.Id.ToString()];
                        if (result == null)
                        {
                            var resolveRequest = new ResolveRequest(service, e.ComponentRegistration, p);
                            result = c.ResolveComponent(resolveRequest);
                            session[e.ComponentRegistration.Id.ToString()] = result;
                        }
                    }

                    return result;
                })
                .As(service)
                .InstancePerLifetimeScope()
                .ExternallyOwned()
                .CreateRegistration());
        }
    });
```

Passing the `ResolveRequest` to the `DelegateActivator` would make this situation easier to manage because you would know which of the original services was being resolved. I wouldn't want to break the existing API though, because most of the time having access to the `IComponentContext` and parameters is all that is required. When building such delegates you usually have an interface that the limit type implements in mind as the `Service`, and tend to resolve dependencies using a particular `Service` as well.

Changing the `ActivateInstance` method on `IInstanceActivator` to accept a `ResolveRequest` would probably be manageable because that is really only intended to be called internally. Changing the signature of the delegate from `Func<IComponentContext, IEnumerable<Parameter>, object>` to `Func<ResolveRequest, object>` would not be desirable because it would break every existing delegate based registration and the `IComponentContext` and parameters would be an extra step away on the `ResolveRequest`.

If we want to cater for this scenario I would add a `RequestResolveDelegateActivator` and allow for the explicit opting in of the `Func<ResolveRequest, object>` signature. This would of course still need to be combined with changing the `ActivateInstance` method to accept the `ResolveRequest`. This might not be worthwhile if there are very few cases where this would be beneficial and workarounds are available.

**Comment by alistairjevans** [comment]
Thanks for the detail @alexmg. 

I think I'll file a PR with your suggested looping approach, then continue going through the integration projects and see if there are any really horrible workarounds that justify the core change.

I'm reluctant to add yet more complexity to Autofac core just to reduce the number of workarounds in older integrations; it's not like I'm finding this in Autofac.Extensions.DependencyInjection.

I just did a PR on the Autofac.Mef repo where I encountered the same problem, but the workarounds were less painful.

Finally, is it just me that is filled with horror at the idea of storing DI objects in the user's session block?
